Fix Builds

mikemhenry opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

mikemhenry committed:
add some swap

mikemhenry committed:
MNT: Re-rendered with conda-build 3.25.0, conda-smithy 3.24.1, and conda-forge-pinning 2023.08.10.11.52.27

mikemhenry committed:
clean up diskspace and only build python 3.9+

mikemhenry committed:
MNT: Re-rendered with conda-build 3.25.0, conda-smithy 3.24.1, and conda-forge-pinning 2023.08.10.11.52.27

mikemhenry committed:
need to use my fork on conda-smithy so I can don't have to re-write this each time

jakirkham commented on recipe/meta.yaml:
```diff
@@ -42,7 +42,7 @@ source:
 build:
   skip: true  # [not linux]
-  number: 0
+  number: 1
```
```suggestion
  number: 1
  skip: true  # [py<39]
```

jakirkham commented on recipe/meta.yaml:
```diff
@@ -76,7 +76,7 @@ requirements:
     - nanoflann
     - parallel-hashmap
-    - python
+    - python >=3.9
```
```suggestion
    - python
```

jakirkham commented on recipe/meta.yaml:
```diff
@@ -87,7 +87,7 @@ requirements:
     - numpy
     - psutil
-    - python
+    - python >=3.9
```
```suggestion
    - python
```

jakirkham reviewed (commented):
Thanks Mike! 🙏

Added a couple suggestions below on constraining `python` version used. Would suggest re-rendering with those changes

mikemhenry committed:
Apply suggestions from code review

Co-authored-by: jakirkham <jakirkham@gmail.com>

mikemhenry commented:
@conda-forge-admin, please rerender

mikemhenry commented:
Thanks @jakirkham I am chipping away at this in between meetings today and forgot we use `skip` (side note, when is conda-forge dropping python 3.8? It has fallen out of NEP-29)

Commit:
MNT: Re-rendered with conda-build 3.26.0, conda-smithy 3.24.1, and conda-forge-pinning 2023.08.10.11.52.27

jakirkham commented:
No worries

We've not set a date yet. Though likely around the same time [Python 3.12 comes out]( https://peps.python.org/pep-0693/#schedule ). So likely early October

mikemhenry committed:
Add swap back in by hand, really need to test the PR

mikemhenry commented:
Okay build is good, I think since this will be the first one, I want to try and add python 3.11 and osx as well...

jakirkham commented:
Can merge PR ( https://github.com/conda-forge/dgl-feedstock/issues/2 ) into this one and re-render (should also resolve conflicts) to enable Python 3.11

mikemhenry commented:
I tested my conda-smithy PR locally and it looks like I've got a bug. I will get the python 3.11 pr merged in + try osx builds (I am not sure if build locally can handle osx builds on linux, I think I need a mac sdk... will have to remember how I did it in the past)

mikemhenry committed:
python 3.11 migration

mikemhenry committed:
give osx a try

mikemhenry committed:
swap added by hand

mikemhenry committed:
fix error with clang

hmacdope commented:
@mikemhenry  libxsmm errors (`error: use of undeclared identifier 'aligned_alloc'`) look like either needed to compile as C++17 or include `cstdlib.h`

mikemhenry committed:
Add back in c++17 flag for non-cuda builds

mikemhenry committed:
just one -

jakirkham commented on recipe/build.sh:
```diff
@@ -3,17 +3,19 @@ set -euxo pipefail
 else
   CUDA_CMAKE_OPTIONS=""
   USE_CUDA=OFF
+  CXXFLAGS="${CXXFLAGS} --std=c++17 "
```
Think it is only one `-`. Also we may want to specify C11

```suggestion
  CFLAGS="${CFLAGS} -std=c11 "
  CXXFLAGS="${CXXFLAGS} -std=c++17 "
```

mikemhenry replied:
@jakirkham I think you posted this RIGHT as I found that mistake :laughing: 
![image](https://github.com/conda-forge/dgl-feedstock/assets/11765982/d9df998a-bf97-4057-9065-a70f226fe1e7)

I am going to see if CI will pass with 17, but 11 might be better. I tired running build local on my linux box and even with the sdk `OSX_SDK_DIR=/home/mmh/Downloads/MacOSX10.9.sdk` I get 

```
Mamba failed to solve:
 - liblapacke 3.9.* *netlib
 - dmlc 0.5.*
 - llvm-openmp >=15.0.7
 - pytorch 2.0.*
 - dlpack 0.7.*
 - metis 5.1.1.*
 - libcxx >=15.0.7
 - libblas 3.9.* *netlib
 - nanoflann
 - pytorch * *cpu*
 - _openmp_mutex >=4.5
 - python 3.9.* *_cpython
 - parallel-hashmap
 - _openmp_mutex * *_llvm
 - cython <3
 - gklib 5.1.1.*
 - liblapack 3.9.* *netlib
 - setuptools
 - libcblas 3.9.* *netlib

with channels:

The reported errors are:
- Encountered problems while solving:
-   - nothing provides requested _openmp_mutex >=4.5
-   - nothing provides requested _openmp_mutex * *_llvm
```

So I don't think I can build an osx package on a linux box

mikemhenry replied:
Getting the same error with 17, I will try 11 but I think I might need to bump the osx sdk so it can use the newer `cstdlib.h`

mikemhenry replied:
Oh and I will try it as a `CFLAGS` first!

mikemhenry committed:
try c17 and c++17

mikemhenry committed:
see if turning off libxsmm and openmp works

mikemhenry committed:
that made things worse

mikemhenry committed:
turn off libxsmm

mikemhenry commented:
Given the error comes in here: `/Users/runner/miniforge3/conda-bld/dgl_1692052195946/work/src/array/cpu/spmm_blocking_libxsmm.h:478:44: error: use of undeclared identifier 'aligned_alloc'` I will see if I can turn off libxsmm, and if that works, I will make it only off for osx

mikemhenry commented:
@hmacdope there is a `#include <unistd.h>` (but looking at that, it doesn't do mem alloc) so should I add a `#include <cstdlib.h>`?

jakirkham commented:
Does the library know how to use [`posix_memalign`]( https://pubs.opengroup.org/onlinepubs/9699919799/functions/posix_memalign.html )?

Looking at [`aligned_alloc` & `posix_memalign`]( https://linux.die.net/man/3/aligned_alloc ), it seems like one could write an `aligned_alloc` wrapper of `posix_memalign`. For example

```C
void* memalign(size_t alignment, size_t size) {
    void* memptr = NULL;
    int err = 0;

    err = posix_memalign(&memptr, alignment, size);
    if (err == 0) {
        return memptr;
    }
    else {
        return NULL;
    }
}
```

That said, Idk how useful this would be. It probably depends on what is using it

jakirkham commented:
FWIW Boost also implements [`aligned_alloc`]( https://www.boost.org/doc/libs/1_83_0/doc/html/align/reference.html#align.reference.functions.aligned_alloc ). Idk if it is worth using Boost just for this though

mikemhenry commented:
So I think that not building `libxsmm` worked but I am getting these errors now: 

```
['  ERROR (dgl,lib/python3.9/site-packages/dgl/dgl_sparse/libdgl_sparse_pytorch_2.0.0.post101.dylib): $RPATH/libdgl.dylib not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  ERROR (dgl,lib/python3.9/site-packages/dgl/_ffi/_cy3/core.cpython-39-darwin.so): $RPATH/libdgl.dylib not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?']
```

RE: using boost, using a wrapper, etc

I rather work with upstream on this if they think it is needed, looking here https://docs.dgl.ai/en/1.1.x/install/index.html#macos they suggest turning off `libxsmm` when building on OSX. It does seem like only `libxsmm` is doing the aligned memory stuff, and I am sure it is needed to make things as fast as they can be, but if you are running dgl on osx, I doubt you are looking for "production" performance since you can't use a GPU for acceleration

mikemhenry commented:
I think I just need to update this line https://github.com/conda-forge/dgl-feedstock/blob/main/recipe/build.sh#L61 to also work on osx now

mikemhenry committed:
use  to fix linking issues

mikemhenry committed:
had two ..

mikemhenry commented:
I want to make sure the linux builds don't throw the same linking errors to make sure things are good. If it builds okay, I will set it up so that OSX skips building `libxsmm`

mikemhenry commented:
Looks like everything is building, I will get these osx builds uploaded to my conda channel and then see if I can find some testers

mikemhenry commented:
And of course I forgot to tell conda-forge to upload artifacts :upside_down_face:

mikemhenry committed:
upload artifacts

mikemhenry committed:
just do osx builds to save CI resources

mikemhenry commented:
Looks like CI stopped, going to open and close to try and kick it off

mikemhenry committed:
bump ci

mikemhenry committed:
MNT: Re-rendered with conda-build 3.25.0, conda-smithy 3.24.1, and conda-forge-pinning 2023.08.15.14.20.10

mikemhenry committed:
do all the builds now

hmacdope reviewed (approved):
Builds seem to be working on OSX now @mikemhenry! Great work. LGTM.

jakirkham commented on recipe/build.sh:
```diff
@@ -3,17 +3,27 @@ set -euxo pipefail
 if [ ${cuda_compiler_version} != "None" ]; then
   CUDA_CMAKE_OPTIONS="-DCMAKE_CUDA_COMPILER=${CUDA_HOME}/bin/nvcc -DCMAKE_CUDA_HOST_COMPILER=${CXX} -DCUDA_ARCH_NAME=All "
   USE_CUDA=ON
+  # Remove -std=c++17 from CXXFLAGS for compatibility with nvcc
+  export CXXFLAGS="$(echo $CXXFLAGS | sed -e 's/ -std=[^ ]*//')"
+  export CFLAGS="$(echo $CFLAGS | sed -e 's/ -mtune=[^ ]*//')"
```
Was the `-mtune` change intended?

Also do we want to remove `std` from `CFLAGS` if present?

mikemhenry replied:
I can experiment to see what happens, @hmacdope added them but I can test that locally.

jakirkham commented on conda-forge.yml:
```diff
@@ -4,3 +4,6 @@ github:
 conda_forge_output_validation: true
+azure:
+  free_disk_space: True
+  upload_packages: True
```
This set to `True` by default. Do we need it set here?

mikemhenry replied:
:upside_down_face: This was me thinking that was the artifact upload, oops!

jakirkham commented on conda-forge.yml:
```diff
@@ -4,3 +4,6 @@ github:
 conda_build:
   error_overlinking: true
 conda_forge_output_validation: true
+azure:
+  free_disk_space: True
+  upload_packages: True
```
Somewhat tangential, but would suggest sorting the top-level keys alphabetical. This is what the migration/update bot does. As a result if these are sorted alphabetically, it will avoid unneeded no-op migration PRs

mikemhenry replied:
Will do!

jakirkham commented on recipe/build.sh:
```diff
@@ -3,17 +3,27 @@ set -euxo pipefail
 else
   CUDA_CMAKE_OPTIONS=""
   USE_CUDA=OFF
+  CFLAGS="${CFLAGS} -std=c17 "
+  CXXFLAGS="${CXXFLAGS} -std=c++17 "
```
Would suggest dropping the trailing space

```suggestion
  CFLAGS="${CFLAGS} -std=c17"
  CXXFLAGS="${CXXFLAGS} -std=c++17"
```

mikemhenry replied:
I am actually not sure if these are needed anymore, if it compiles without them, I will drop them.

jakirkham commented on recipe/meta.yaml:
```diff
@@ -41,8 +42,9 @@ source:
 build:
-  skip: true  # [not linux]
-  number: 0
+  skip: true  # [win]
```
```suggestion
```

jakirkham commented on recipe/meta.yaml:
```diff
@@ -41,8 +42,9 @@ source:
-  number: 0
+  skip: true  # [win]
+  number: 1
+  skip: true  # [py<39]
```
Might be worth grouping the `skip`s together or possibly putting them all in one line like so

```suggestion
  skip: true  # [win or py<39]
```

jakirkham commented on recipe/meta.yaml:
```diff
@@ -11,6 +11,7 @@ source:
     sha256: 71c8ba7777ffdca37638bc796524d42c51006fad385967fa140235b60aac0252
     patches:
       - conda-build.patch
+      - fix_clang_errors.patch
```
Would suggest sending these upstream if not done already. Should allow these to be dropped in the future

hmacdope replied:
@jakirkham I am in the process of getting ready to upstream these. :) Hopefully will go smoothly.

mikemhenry replied:
Yes! So our plan is upstream the cmake patch, we will need patches for 1.1.2 as well. The clang patch is a back port from upstream, so when we build the next version we will not need it.

jakirkham reviewed (commented):
Thanks Mike! 🙏

Had a few suggestions below

Also is this now re-rendering with the conda-smithy PR?

mikemhenry committed:
applied sugestions from code review + testing mtune

mikemhenry committed:
added swap by hand (need to fix that pr grrr)

mikemhenry committed:
I think we need to remove std  C++17 for nvcc

mikemhenry commented:
Okay looks like we can keep mtune but need to strip the std flag, will wait for build to finish to confirm.

mikemhenry committed:
add swap using new conda-smithy

mikemhenry committed:
lets see if we can speed these builds up without OOMing ourselves

mikemhenry commented:
Once these builds finish, I just want to upload them to my anaconda channel and see if someone other than me can use them to run the dgl tutorial or something, especially on OSX

jakirkham commented on conda-forge.yml:
```diff
@@ -1,6 +1,10 @@
-  tooling_branch_name: main
+azure:
+  free_disk_space: true
+  store_build_artifacts: true
```
Would just remember to revert this and re-render after testing is done

mikemhenry replied:
I will have to re-add the swap by hand (not a big deal), I ran `conda smithy rerender` and it didn't complain about the extra keys (maybe I should make a PR about that, since it would catch typos), so is there any danger leaving it in?

jakirkham replied:
Yeah was more meaning dropping the `store_build_artifacts` field

mikemhenry replied:
Oh! I kinda like using that so I can test the builds, does it chew through a lot of resources?

jakirkham replied:
It can yeah. Generally it is better to enable for debugging and then turn off. It's pretty simple to readd when needed though

mikemhenry replied:
Good to know, thanks!

jakirkham reviewed (commented):
Sounds good. Left one note for follow up after testing

mikemhenry committed:
removed new arg for swap file, kept swap file because it is needed for build

mikemhenry commented:
I think we will need to do a patch for the way python loads `libdgl_sparse_pytorch_2.0.0`.

So with the way we build `dgl` on conda-forge, we only build against a single version of pytorch (or if we do more, it will be part of a matrix, not one build supporting multiple pytorches, so I think we can change the logic here:

https://github.com/dmlc/dgl/blob/1.1.0/python/dgl/sparse/__init__.py#L19

To just grab the file with a glob, and throw an error if there is more than one match. 

I believe this is needed since I am seeing this issue:

The GPU build on linux makes this file `libdgl_sparse_pytorch_2.0.0.post200.so` and the logic for finding that version is like this `python -c 'import torch; print(torch.__version__.split("+", maxsplit=1)[0])'` which reports `2.0.0.post200` for the GPU build.

The CPU build on linux makes this file `libdgl_sparse_pytorch_2.0.0.so` and version reported with `python -c 'import torch; print(torch.__version__.split("+", maxsplit=1)[0])'` is  `2.0.0.post101` which doesn't match the version of the object file, so it barfs like this:

`E   FileNotFoundError: Cannot find DGL C++ sparse library at /home/mmh/micromamba/envs/espaloma-new-dgl-CPU/lib/python3.11/site-packages/dgl/dgl_sparse/libdgl_sparse_pytorch_2.0.0.post101.so`

mikemhenry commented:
Some more info here: https://github.com/dmlc/dgl/issues/6004

mikemhenry commented:
Also we need to get a test added that triggers this so we can catch it in CI.

mikemhenry commented:
Looks like this is enough `import dgl.sparse`

mikemhenry committed:
add test to make sure we can load dgl sparse

mikemhenry commented:
Looks like only a problem on python 3.9 and 3.11 on linux-cpu, OSX seems okay

mikemhenry commented:
Really don't understand how this is passing that import check, but failing in CI here: https://github.com/choderalab/espaloma/pull/179

Perhaps there is an issue with a pin or something

mikemhenry commented:
Okay so from the build log:
```
2023-08-17T20:14:36.6987193Z [100%] Linking CXX shared library libdgl_sparse_pytorch_2.0.0.post101.so
2023-08-17T20:14:36.8550750Z [100%] Built target dgl_sparse_pytorch_2.0.0.post101
2023-08-17T20:14:36.8956409Z ‘libdgl_sparse_pytorch_2.0.0.post101.so’ -> ‘$SRC_DIR/build/dgl_sparse/libdgl_sparse_pytorch_2.0.0.post101.so’
```

But when I download the package:

```
/home/mmh/micromamba/envs/dgl-beta-CPU-py39/lib/python3.9/site-packages/dgl/dgl_sparse/libdgl_sparse_pytorch_2.0.0.so
```

The file name is different?!

mikemhenry committed:
add a patch to make it more robust when looking for libdgl_sparse_pytorch shared lib

mikemhenry commented:
Okay I added a patch that makes the logic for finding libdgl_sparse_pytorch more robust, but I am still not sure if there is a real issue or not. At least locally for me this makes things work, but I am worried that I am missing something...

mikemhenry commented:
espaloma builds are passing :tada:

mikemhenry commented:
Going to turn on auto merge :smile:
